projects/m2/003-remove-outliers/solutions/andreaattardo/solution_remove-outliers.c

```diff
@@ -0,0 +1,155 @@
+//remove-outliers.c
+// ---------------
+//When analysing data collected as part of a science experiment it may be desirable to remove the most extreme values before performing other calculations.
+//Write a function that takes a list of values and an non - negative integer, n, as its parameters.The function should create a new copy of the list with the n largest
+//elementsand the n smallest elements removed.Then it should return the new copy of the list as the function�s only result.
+
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <stdbool.h>
+
+//CONSTANTS
+//----------
+#define USRINPUT_LEN 50 //usr input stream length
+
+
+//FUNCTIONS DECLARATIONS
+// --------------------
+int* sortIntArray(int* intArray, int arrayElements);
+int getInt();
+int* initDynamicArray(int arrayElements);
+void printIntArray(int* intArray, int arrayElements);
+int* removeOutliers(int* intArray, int arrayElements, int nOutliers);
+
+
+//MAIN
+int main() {
+	int* sciData, *cuttedSciData;
+	int sciDataElements = 0, cuttedSciDataElements = 0, nOutliers = 0;
+
+	printf("This program remove an user defined numbers of extreme values from a list of integers.\n\n");
+	printf("Creating a list of integers inserted one by one,\n");
+	printf("how many elements do you want to add?: ");
+	sciDataElements = getInt();
+
+	printf("Insert elements:\n");
+	sciData = initDynamicArray(sciDataElements);
+
+	sortIntArray(sciData, sciDataElements);
+
+	printf("How many extreme values do you want to remove from the above list?\n");
+	nOutliers = getInt();
+
+	cuttedSciDataElements = (sciDataElements - (nOutliers * 2));
+	cuttedSciData = removeOutliers(sciData, sciDataElements, nOutliers);
+
+	printf("The new list with %d values on each extreme removed:\n", nOutliers);
+	printIntArray(cuttedSciData, cuttedSciDataElements);
+
+	return 0;
+}
+
+
+
+//FUNCTIONS DEFINITIONS
+//-----------------------
+
+/// Function: getInt
+/// 
+/// return an integer from stdin
+/// <returns></returns>
+int getInt() {
+	char* usrInput[USRINPUT_LEN * sizeof(char)];
+	int usrInt = 0;
+
+	fgets(usrInput, USRINPUT_LEN, stdin);
+	sscanf_s(usrInput, "%d", &usrInt);
+
+	return usrInt;
+}
+
+///Function: initDynamicArray
+/// 
+/// Initialize an Array of integers
+/// manually allocated in memory
+/// and add values entered by the usr
+/// across the lenght of the array.
+/// Return an allocated array of integers.
+int* initDynamicArray(int arrayElements) {
+
+	int usrNum = 0;
+	int numCount = 0;
+
+	int* intArray = malloc(arrayElements * sizeof(int));
```

Comment:
un piccolo trick per usare correttamente la malloc():

se in futuro (cioè ipotizzando che stiamo parlando di un software reale con un suo ciclo di vita) questa variabile dovesse cambiare tipo, ad esempio diventando un long int \*, allora bisognerebbe ricordarsi di modificare la malloc, aggiornando il sizeof() con il nuovo tipo all'interno.
Un trick che puoi usare è quello di fare sizeof(\*intArray) in questo modo dici al compilatore di usare il size dell'oggetto puntato. Se in futuro dovesse cambiare in char * oppure in long int* , la sizeof() sarebbe automaticamente corretta senza dover intervenire manualmente.

Reply:
Bello questo fatto, io pensavo che dato che intArray fosse ancora da inizializzare, il compilatore non sapesse che il tipo di byte al momento fosse int.
Il programma funziona ma il compilatore (sto usando visual studio) mi dà questi 2 warning, che significano?
Buffer overrun while writing to 'intArray'	
Dereferencing NULL pointer 'intArray'.